### PR TITLE
Feature/#34 twitter

### PR DESCRIPTION
**Pull request by benetherington**
Not too much here, just a quick Twitter integration. The only thing specific to the bot is that the returned data is in a similar format to other things that get pushed into the database.

I'm not sure if I want to have separate tables for all of the TWSF data sources (tweets, DMs, emails, discord messages), or if they should all go into one table with a type column. Each strategy has pros and cons in my mind. I'm leaning towards the latter at the moment, though.

### CODE REVIEW

**Comment by fonji** [comment]
> I'm not sure if I want to have separate tables for all of the TWSF data sources (tweets, DMs, emails, discord messages), or if they should all go into one table with a type column. Each strategy has pros and cons in my mind. I'm leaning towards the latter at the moment, though.

I also prefer the latter.

File: integrations/twitter-v1.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+const fetchTwsfDirectMessages = async (cursor = undefined) => {
+    // Fetch direct message events
+    const {events} = await client.get('direct_messages/events/list.json', {
+        cursor,
+    });
+    console.log(`Found ${events.length} direct messsages.`);
+
+    // Filter messages. I'm not sure another type is ever provided, but can't
+    // hurt to be safe.
+    const directMessages = events.filter(
```
Comment (fonji) [inline integrations/twitter-v1.js:47]: merge all the filters
```js
event.type == 'message_create' && sender_id !== TOMUserId && ...
```
Do not hesitate to create a named function to make it pretty

File: integrations/twitter-v1.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        `${hashtagMessages.length} of them are #ThisWeekSF direct messsages.`,
+    );
+
+    // Extract the data we want
+    const formattedPromises = hashtagMessages.map(formatDirectMessage);
+    const formattedMessages = await Promise.all(formattedPromises);
+
+    // Check for another page
+    if (events.nextCursor) {
+        // Recurse, passing along the cursor
```
Comment (fonji) [inline integrations/twitter-v1.js:72]: I tend to prefer a while loop which call the function that makes the call, than having recursion.
Or having a simple recursive function that just concats all the messages, before any filtering or anything.
Here, the recursion appears by surprise at the end of a long function. So we don't see it coming, and this function has too many responsabilities.
An ideal function should have one job -- take this with a grain of salt though.

File: integrations/twitter-v1.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+};
+
+module.exports.storeNewTwsfDirectMessages = async () => {
+    console.log('Storing #ThisWeekSF direct messages...');
+
+    // Get new DMs
+    const twsfDms = await fetchTwsfDirectMessages();
+
+    // Store new DMs
+    const storeagePromises = twsfDms.map((dm) => addNewTwsfGuess(dm));
```
Comment (fonji) [inline integrations/twitter-v1.js:90]: storEage?

File: integrations/twitter-v2.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
+try {
```
Comment (fonji) [inline integrations/twitter-v2.js:1]: wait what?
Why did I read all of the v1 then? :smiling_face_with_tear: 

Stopping here, gotta sleep. But what should I read?

File: integrations/twitter-v1.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+    const {name: twitterName, screen_name: twitterUsername} = authorData;
+
+    return {
+        author: {twitterId, twitterName, twitterUsername},
+        guess: {type: 'twitter dm', twitterDmId, text},
+    };
+};
+
+const fetchTwsfDirectMessages = async (cursor = undefined) => {
```
[the remark this answers is not in the document]
Comment (benetherington) [reply integrations/twitter-v1.js:38]: `mv ./integrations ./twsf` and `mv ./interface ./title-suggestion`? I can get behind that.

File: integrations/twitter-v1.js (hunk 2 above)
Comment (benetherington) [reply integrations/twitter-v1.js:72]: Yes, looping is better here. I jump to recursion right away, and it has bitten me in the ass more times than I can count.

File: integrations/twitter-v1.js (hunk 3 above)
Comment (benetherington) [reply integrations/twitter-v1.js:90]: 🙄

File: integrations/twitter-v2.js (hunk 4 above)
Comment (benetherington) [reply integrations/twitter-v2.js:1]: Twitter has two API versions. I probably don't need to migrate to v2 yet, but I started there before I realized that API doesn't yet support DMs. I'm fetching hashtag tweets from API v1, DMs from API v2.

Comment (fonji) [reply integrations/twitter-v2.js:1]: Oh OK. TLDR: twitter API sucks? :grin:
Raises the question if we need to reflect that on our side. Maybe just `integration/twitter` and some functions use the `v1` and others the `v2`? I'll know more after reading this file.

Comment (benetherington) [reply integrations/twitter-v2.js:1]: Kk. Another option is to just use V1 exclusively. I'm sure they're not going to deprecate it any time soon if V2 isn't feature complete.

**Comment by fonji** [comment]
Good job :+1:

**Pull request merged**